File: csharp/CsSearch/CsSearch/SearchSettings.cs
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace CsSearch
{
	public class SearchSettings
	{
		public bool ArchivesOnly { get; set; }
		public bool Debug { get; set; }
		public bool ExcludeHidden { get; set; }
		public bool FirstMatch { get; set; }
		public ISet<string> InArchiveExtensions { get; private set; }
		public ISet<Regex> InArchiveFilePatterns { get; private set; }
		public ISet<Regex> InDirPatterns { get; private set; }
		public ISet<string> InExtensions { get; private set; }
		public ISet<Regex> InFilePatterns { get; private set; }
		public ISet<FileType> InFileTypes { get; private set; }
		public ISet<Regex> InLinesAfterPatterns { get; private set; }
		public ISet<Regex> InLinesBeforePatterns { get; private set; }
		public int LinesAfter { get; set; }
		public ISet<Regex> LinesAfterToPatterns { get; private set; }
		public ISet<Regex> LinesAfterUntilPatterns { get; private set; }
		public int LinesBefore { get; set; }
		public bool ListDirs { get; set; }
		public bool ListFiles { get; set; }
		public bool ListLines { get; set; }
		public int MaxLineLength { get; set; }
		public bool MultiLineSearch { get; set; }
		public ISet<string> OutArchiveExtensions { get; private set; }
		public ISet<Regex> OutArchiveFilePatterns { get; private set; }
		public ISet<Regex> OutDirPatterns { get; private set; }
		public ISet<string> OutExtensions { get; private set; }
		public ISet<Regex> OutFilePatterns { get; private set; }
		public ISet<FileType> OutFileTypes { get; private set; }
		public ISet<Regex> OutLinesAfterPatterns { get; private set; }
		public ISet<Regex> OutLinesBeforePatterns { get; private set; }
		public bool PrintResults { get; set; }
		public bool PrintUsage { get; set; }
		public bool PrintVersion { get; set; }
		public bool Recursive { get; set; }
		public bool SearchArchives { get; set; }
		public ISet<Regex> SearchPatterns { get; private set; }
		public string StartPath { get; set; }
		public string TextFileEncoding { get; set; }
		public bool UniqueLines { get; set; } = false;
		public bool Verbose { get; set; }

		public SearchSettings()
		{
			ArchivesOnly = false;
			Debug = false;
			ExcludeHidden = true;
			FirstMatch = false;
			InArchiveExtensions = new HashSet<string>();
			InArchiveFilePatterns = new HashSet<Regex>();
			InDirPatterns = new HashSet<Regex>();
			InExtensions = new HashSet<string>();
			InFilePatterns = new HashSet<Regex>();
			InFileTypes = new HashSet<FileType>();
			InLinesAfterPatterns = new HashSet<Regex>();
			InLinesBeforePatterns = new HashSet<Regex>();
			LinesAfter = 0;
			LinesAfterToPatterns = new HashSet<Regex>();
			LinesAfterUntilPatterns = new HashSet<Regex>();
			LinesBefore = 0;
			ListDirs = false;
			ListFiles = false;
			ListLines = false;
			MaxLineLength = 150;
			MultiLineSearch = false;
			OutArchiveExtensions = new HashSet<string>();
			OutArchiveFilePatterns = new HashSet<Regex>();
			OutDirPatterns = new HashSet<Regex>();
			OutExtensions = new HashSet<string>();
			OutFilePatterns = new HashSet<Regex>();
			OutFileTypes = new HashSet<FileType>();
			OutLinesAfterPatterns = new HashSet<Regex>();
			OutLinesBeforePatterns = new HashSet<Regex>();
			PrintResults = false;
			PrintUsage = false;
			PrintVersion = false;
			Recursive = true;
			SearchArchives = false;
			SearchPatterns = new HashSet<Regex>();
			TextFileEncoding = "utf-8";
			UniqueLines = false;
			Verbose = false;
		}

		private static void AddExtension(ISet<string> set, string extList)
		{
			var exts = extList.Split(new[] { ',' });
			foreach (var x in exts)
			{
				var ext = x;
				if (!ext.StartsWith("."))
					ext = "." + ext;
				set.Add(ext.ToLowerInvariant());
			}
		}

		public void AddInExtension(string ext)
		{
			AddExtension(InExtensions, ext);
		}

		public void AddOutExtension(string ext)
		{
			AddExtension(OutExtensions, ext);
		}

		private static void AddPattern(ISet<Regex> set, string pattern)
		{
			set.Add(new Regex(pattern));
		}

		public void AddInDirPattern(string pattern)
		{
			AddPattern(InDirPatterns, pattern);
		}

		public void AddOutDirPattern(string pattern)
		{
			AddPattern(OutDirPatterns, pattern);
		}

		public void AddInFilePattern(string pattern)
		{
			AddPattern(InFilePatterns, pattern);
		}

		public void AddOutFilePattern(string pattern)
		{
			AddPattern(OutFilePatterns, pattern);
		}

		public void AddInArchiveExtension(string ext)
		{
			AddExtension(InArchiveExtensions, ext);
		}

		public void AddOutArchiveExtension(string ext)
		{
			AddExtension(OutArchiveExtensions, ext);
		}

		public void AddInArchiveFilePattern(string pattern)
		{
			AddPattern(InArchiveFilePatterns, pattern);
		}

		public void AddOutArchiveFilePattern(string pattern)
		{
			AddPattern(OutArchiveFilePatterns, pattern);
		}

		public void AddInLinesAfterPattern(string pattern)
		{
			AddPattern(InLinesAfterPatterns, pattern);
		}

		public void AddOutLinesAfterPattern(string pattern)
		{
			AddPattern(OutLinesAfterPatterns, pattern);
		}

		public void AddInLinesBeforePattern(string pattern)
		{
			AddPattern(InLinesBeforePatterns, pattern);
		}

		public void AddOutLinesBeforePattern(string pattern)
		{
			AddPattern(OutLinesBeforePatterns, pattern);
		}

		public void AddLinesAfterToPattern(string pattern)
		{
			AddPattern(LinesAfterToPatterns, pattern);
		}

		public void AddLinesAfterUntilPattern(string pattern)
		{
			AddPattern(LinesAfterUntilPatterns, pattern);
		}

		public void AddSearchPattern(string pattern)
		{
			AddPattern(SearchPatterns, pattern);
		}

		private static void AddFileType(ISet<FileType> set, string typeNameList)
		{
			var typeNames = typeNameList.Split(new[] { ',' });
			foreach (var t in typeNames)
			{
				set.Add(FileTypes.FromName(t));
			}
		}

		public void AddInFileType(string typeName)
		{
			AddFileType(InFileTypes, typeName);
		}

		public void AddOutFileType(string typeName)
		{
			AddFileType(OutFileTypes, typeName);
		}

		public void SetArchivesOnly(bool archivesOnly)
		{
			ArchivesOnly = archivesOnly;
			if (archivesOnly)
				SearchArchives = true;
		}

		public void SetDebug(bool debug)
		{
			Debug = debug;
			if (debug)
				Verbose = true;
		}

		protected static string EnumerableToString<T>(IEnumerable<T> enumerable)
		{
			var sb = new StringBuilder("[");
			var elemCount = 0;
			foreach (var x in enumerable)
			{
				var t = x.GetType();
				if (elemCount > 0)
					sb.Append(", ");
				if (t == typeof(string))
					sb.Append("\"");
				sb.Append(x);
				if (t == typeof(string))
					sb.Append("\"");
				elemCount++;
			}
			sb.Append("]");
			return sb.ToString();
		}

		public override string ToString()
		{
			var sb = new StringBuilder("SearchSettings(");
			sb.Append("ArchivesOnly: " + ArchivesOnly);
			sb.Append(", Debug: " + Debug);
			sb.Append(", ExcludeHidden: " + ExcludeHidden);
			sb.Append(", FirstMatch: " + FirstMatch);
			sb.Append(", InArchiveExtensions: " + EnumerableToString(InArchiveExtensions));
			sb.Append(", InArchiveFilePatterns: " + EnumerableToString(InArchiveFilePatterns));
			sb.Append(", InDirPatterns: " + EnumerableToString(InDirPatterns));
			sb.Append(", InExtensions: " + EnumerableToString(InExtensions));
			sb.Append(", InFilePatterns: " + EnumerableToString(InFilePatterns));
			sb.Append(", InLinesAfterPatterns: " + EnumerableToString(InLinesAfterPatterns));
			sb.Append(", InLinesBeforePatterns: " + EnumerableToString(InLinesBeforePatterns));
			sb.Append(", LinesAfter: " + LinesAfter);
			sb.Append(", LinesAfterToPatterns: " + EnumerableToString(LinesAfterToPatterns));
			sb.Append(", LinesAfterUntilPatterns: " + EnumerableToString(LinesAfterUntilPatterns));
			sb.Append(", LinesBefore: " + LinesBefore);
			sb.Append(", ListDirs: " + ListDirs);
			sb.Append(", ListFiles: " + ListFiles);
			sb.Append(", ListLines: " + ListLines);
			sb.Append(", MaxLineLength: " + MaxLineLength);
			sb.Append(", MultiLineSearch: " + MultiLineSearch);
			sb.Append(", OutArchiveExtensions: " + EnumerableToString(OutArchiveExtensions));
			sb.Append(", OutArchiveFilePatterns: " + EnumerableToString(OutArchiveFilePatterns));
			sb.Append(", OutDirPatterns: " + EnumerableToString(OutDirPatterns));
			sb.Append(", OutExtensions: " + EnumerableToString(OutExtensions));
			sb.Append(", OutFilePatterns: " + EnumerableToString(OutFilePatterns));
			sb.Append(", OutLinesAfterPatterns: " + EnumerableToString(OutLinesAfterPatterns));
			sb.Append(", OutLinesBeforePatterns: " + EnumerableToString(OutLinesBeforePatterns));
			sb.Append(", PrintResults: " + PrintResults);
			sb.Append(", PrintUsage: " + PrintUsage);
			sb.Append(", PrintVersion: " + PrintVersion);
			sb.Append(", Recursive: " + Recursive);
			sb.Append(", SearchArchives: " + SearchArchives);
			sb.Append(", SearchPatterns: " + EnumerableToString(SearchPatterns));
			sb.Append(", StartPath: \"" + StartPath + "\"");
			sb.Append(", TextFileEncoding: \"" + TextFileEncoding + "\"");
			sb.Append(", UniqueLines: " + UniqueLines);
			sb.Append(", Verbose: " + Verbose);
			sb.Append(")");
			return sb.ToString();
		}
	}
}

File: php/phpsearch/searchsettings.php
<?php
/**
 * Class SearchSettings
 *
 * @property bool archivesonly
 * @property bool debug
 * @property bool excludehidden
 * @property bool firstmatch
 * @property array in_archiveextensions
 * @property array in_archivefilepatterns
 * @property array in_dirpatterns
 * @property array in_extensions
 * @property array in_filepatterns
 * @property array in_filetypes
 * @property array in_linesafterpatterns
 * @property array in_linesbeforepatterns
 * @property int linesafter
 * @property array linesaftertopatterns
 * @property array linesafteruntilpatterns
 * @property int linesbefore
 * @property bool listdirs
 * @property bool listfiles
 * @property bool listlines
 * @property int maxlinelength
 * @property bool multilinesearch
 * @property array out_archiveextensions
 * @property array out_archivefilepatterns
 * @property array out_dirpatterns
 * @property array out_extensions
 * @property array out_filepatterns
 * @property array out_filetypes
 * @property array out_linesafterpatterns
 * @property array out_linesbeforepatterns
 * @property bool printresults
 * @property bool printusage
 * @property bool printversion
 * @property bool recursive
 * @property bool searcharchives
 * @property array searchpatterns
 * @property string startpath
 * @property string textfileencoding
 * @property bool uniquelines
 * @property bool verbose
 */
class SearchSettings {
    public $archivesonly = false;
    public $debug = false;
    public $excludehidden = true;
    public $firstmatch = false;
    public $linesafter = 0;
    public $linesbefore = 0;
    public $listdirs = false;
    public $listfiles = false;
    public $listlines = false;
    public $maxlinelength = 150;
    public $multilinesearch = false;
    public $printresults = true;
    public $printusage = false;
    public $printversion = false;
    public $recursive = true;
    public $searcharchives = false;
    public $startpath = NULL;
    public $textfileencoding = 'utf-8';
    public $uniquelines = false;
    public $verbose = false;

    public $in_archiveextensions = array();
    public $in_archivefilepatterns = array();
    public $in_dirpatterns = array();
    public $in_extensions = array();
    public $in_filepatterns = array();
    public $in_filetypes = array();
    public $in_linesafterpatterns = array();
    public $in_linesbeforepatterns = array();
    public $linesaftertopatterns = array();
    public $linesafteruntilpatterns = array();
    public $out_archiveextensions = array();
    public $out_archivefilepatterns = array();
    public $out_dirpatterns = array();
    public $out_extensions = array();
    public $out_filepatterns = array();
    public $out_filetypes = array();
    public $out_linesafterpatterns = array();
    public $out_linesbeforepatterns = array();
    public $searchpatterns = array();

    public function add_exts($ext, &$exts) {
        if (gettype($ext) == 'string') {
            $xs = explode(',', $ext);
            foreach ($xs as $x) {
                $exts[] = $x;
            }
        } elseif (gettype($ext) == 'array') {
            foreach ($ext as $x) {
                $exts[] = $x;
            }
        }
    }

    public function add_filetypes($filetype, &$filetypes) {
        if (gettype($filetype) == 'string') {
            $fts = explode(',', $filetype);
            foreach ($fts as $ft) {
                $filetypes[] = FileTypes::from_name($ft);
            }
        } elseif (gettype($filetype) == 'array') {
            foreach ($filetype as $ft) {
                $filetypes[] = FileTypes::from_name($ft);
            }
        }
    }

    public function add_patterns($pattern, &$patterns) {
        if (gettype($pattern) == 'string') {
            $patterns[] = $pattern;
        } elseif (gettype($pattern) == 'array') {
            foreach ($pattern as $p) {
                $patterns[] = $p;
            }
        }
    }

    public function set_archivesonly(bool $b) {
        $this->archivesonly = $b;
        if ($b) {
            $this->searcharchives = $b;
        }
    }

    public function set_debug(bool $b) {
        $this->debug = $b;
        if ($b) {
            $this->verbose = $b;
        }
    }

    private function arr_to_string(array $arr): string {
        $s = '["' . implode('","', $arr) . '"]';
        return $s;
    }

    private function bool_to_string(bool $b): string {
        return $b ? 'true' : 'false';
    }

    public function __toString(): string {
        $s = 'SearchSettings(';
        $s .= 'archivesonly: ' . $this->bool_to_string($this->archivesonly);
        $s .= ', debug: ' . $this->bool_to_string($this->debug);
        $s .= ', excludehidden: ' . $this->bool_to_string($this->excludehidden);
        $s .= ', firstmatch: ' . $this->bool_to_string($this->firstmatch);
        $s .= ', in_archiveextensions: ' . $this->arr_to_string($this->in_archiveextensions);
        $s .= ', in_archivefilepatterns: ' . $this->arr_to_string($this->in_archivefilepatterns);
        $s .= ', in_dirpatterns: ' . $this->arr_to_string($this->in_dirpatterns);
        $s .= ', in_extensions: ' . $this->arr_to_string($this->in_extensions);
        $s .= ', in_filepatterns: ' . $this->arr_to_string($this->in_filepatterns);
        $s .= ', in_filetypes: ' . $this->arr_to_string($this->in_filetypes);
        $s .= ', in_linesafterpatterns: ' . $this->arr_to_string($this->in_linesafterpatterns);
        $s .= ', in_linesbeforepatterns: ' . $this->arr_to_string($this->in_linesbeforepatterns);
        $s .= ', linesafter: ' . $this->linesafter;
        $s .= ', linesaftertopatterns: ' . $this->arr_to_string($this->linesaftertopatterns);
        $s .= ', linesafteruntilpatterns: ' . $this->arr_to_string($this->linesafteruntilpatterns);
        $s .= ', linesbefore: ' . $this->linesbefore;
        $s .= ', listdirs: ' . $this->bool_to_string($this->listdirs);
        $s .= ', listfiles: ' . $this->bool_to_string($this->listfiles);
        $s .= ', listlines: ' . $this->bool_to_string($this->listlines);
        $s .= ', maxlinelength: ' . $this->maxlinelength;
        $s .= ', multilinesearch: ' . $this->bool_to_string($this->multilinesearch);
        $s .= ', out_archiveextensions: ' . $this->arr_to_string($this->out_archiveextensions);
        $s .= ', out_archivefilepatterns: ' . $this->arr_to_string($this->out_archivefilepatterns);
        $s .= ', out_dirpatterns: ' . $this->arr_to_string($this->out_dirpatterns);
        $s .= ', out_extensions: ' . $this->arr_to_string($this->out_extensions);
        $s .= ', out_filepatterns: ' . $this->arr_to_string($this->out_filepatterns);
        $s .= ', out_filetypes: ' . $this->arr_to_string($this->out_filetypes);
        $s .= ', out_linesafterpatterns: ' . $this->arr_to_string($this->out_linesafterpatterns);
        $s .= ', out_linesbeforepatterns: ' . $this->arr_to_string($this->out_linesbeforepatterns);
        $s .= ', printresults: ' . $this->bool_to_string($this->printresults);
        $s .= ', printusage: ' . $this->bool_to_string($this->printusage);
        $s .= ', printversion: ' . $this->bool_to_string($this->printversion);
        $s .= ', recursive: ' . $this->bool_to_string($this->recursive);
        $s .= ', searcharchives: ' . $this->bool_to_string($this->searcharchives);
        $s .= ', searchpatterns: ' . $this->arr_to_string($this->searchpatterns);
        $s .= ', startpath: "' . $this->startpath . '"';
        $s .= ', textfileencoding: "' . $this->textfileencoding . '"';
        $s .= ', uniquelines: ' . $this->bool_to_string($this->uniquelines);
        $s .= ', verbose: ' . $this->bool_to_string($this->verbose);
        $s .= ')';
        return $s;
    }
}

?>
